File: subprojects/xcfa/cat/src/main/java/hu/bme/mit/theta/cat/dsl/CatVisitor.java
/*
 *  Copyright 2022 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package hu.bme.mit.theta.cat.dsl;

import hu.bme.mit.theta.analysis.algorithm.mcm.MCM;
import hu.bme.mit.theta.analysis.algorithm.mcm.MCMConstraint;
import hu.bme.mit.theta.analysis.algorithm.mcm.MCMRelation;
import hu.bme.mit.theta.analysis.algorithm.mcm.rules.*;
import hu.bme.mit.theta.cat.dsl.gen.CatBaseVisitor;
import hu.bme.mit.theta.cat.dsl.gen.CatParser;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.google.common.base.Preconditions.checkNotNull;
import static hu.bme.mit.theta.cat.dsl.CatDslManager.setupCatAntlr;
import static java.lang.Character.isUpperCase;

public class CatVisitor extends CatBaseVisitor<MCMRelation> {
    private MCM mcm;
    private int ruleNameCnt = 0;
    private final File file;
    private final Map<String, MCMRelation> relations;
    private final Map<String, CatParser.FunctionDefContext> functions;
    private final Map<String, CatParser.ProcDefContext> procedures;
    private final Map<String, MCMRelation> paramAssignment;
    private static final Set<String> unaryRelations = Set.of("W", "R", "F", "U");

    public CatVisitor(final File file) {
        this.relations = new LinkedHashMap<>();
        this.procedures = new LinkedHashMap<>();
        this.functions = new LinkedHashMap<>();
        this.paramAssignment = new LinkedHashMap<>();
        this.file = file;
    }

    public MCM getMcm() {
        return mcm;
    }

    private MCMRelation getOrCreateRelation(final String name, final int arity) {
        if(paramAssignment.containsKey(name)) {
            final MCMRelation mcmRelation = paramAssignment.get(name);
            return mcmRelation;
        }
        if(relations.containsKey(name)) {
            final MCMRelation mcmRelation = relations.get(name);
            return mcmRelation;
        }
        final MCMRelation relation = new MCMRelation(arity, name);
        relations.put(name, relation);
        return relation;
    }

    @Override
    public MCMRelation visitMcm(CatParser.McmContext ctx) {
        this.mcm = new MCM(ctx.name() == null ? "Unnamed" : ctx.name().getText());

        final File file = new File(this.file.getParent() + File.separator + "stdlib.cat");
        visitIncludedFile(file);

        return super.visitMcm(ctx);
    }

    private void visitIncludedFile(File file) {
        if(file.exists() && file.isFile()) {
            try {
                final CatParser.McmContext context = setupCatAntlr(file);
                context.scopeBody().accept(this);
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else throw new RuntimeException(new FileNotFoundException(file.getAbsolutePath()));
    }

    @Override
    public MCMRelation visitIncludeFile(CatParser.IncludeFileContext ctx) {
        final File file = new File(this.file.getParent() + File.separator + ctx.file.getText());
        visitIncludedFile(file);
        return null;
    }

    @Override
    public MCMRelation visitAcyclicDefinition(CatParser.AcyclicDefinitionContext ctx) {
        final MCMRelation rule = ctx.e.accept(this);
        mcm.addConstraint(new MCMConstraint(rule, ctx.negate == null ? MCMConstraint.ConstraintType.ACYCLIC : MCMConstraint.ConstraintType.CYCLIC));
        return null;
    }

    @Override
    public MCMRelation visitIrreflexiveDefinition(CatParser.IrreflexiveDefinitionContext ctx) {
        final MCMRelation rule = ctx.e.accept(this);
        mcm.addConstraint(new MCMConstraint(rule, ctx.negate == null ? MCMConstraint.ConstraintType.IRREFLEXIVE : MCMConstraint.ConstraintType.REFLEXIVE));
        return null;
    }

    @Override
    public MCMRelation visitEmptyDefinition(CatParser.EmptyDefinitionContext ctx) {
        final MCMRelation rule = ctx.e.accept(this);
        mcm.addConstraint(new MCMConstraint(rule, ctx.negate == null ? MCMConstraint.ConstraintType.EMPTY : MCMConstraint.ConstraintType.NOTEMPTY));
        return null;
    }

    @Override
    public MCMRelation visitFunctionDef(CatParser.FunctionDefContext ctx) {
        functions.put(ctx.n.getText(), ctx);
        return null;
    }

    @Override
    public MCMRelation visitProcDef(CatParser.ProcDefContext ctx) {
        procedures.put(ctx.n.getText(), ctx);
        return null;
    }

    @Override
    public MCMRelation visitExprToid(CatParser.ExprToidContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, 2);
        relation.addRule(new Toid(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExprRange(CatParser.ExprRangeContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, 1);
        relation.addRule(new Range(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExprDomain(CatParser.ExprDomainContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, 1);
        relation.addRule(new Domain(rel));
        return relation;
    }

    @Override
    public MCMRelation visitProcCall(CatParser.ProcCallContext ctx) {
        final CatParser.ProcDefContext procDefContext = procedures.get(ctx.NAME().getText());
        checkNotNull(procDefContext, "Procedure with name " + ctx.NAME().getText() + " does not exist.");
        final List<CatParser.ExpressionContext> e = ctx.params;
        final Map<String, MCMRelation> toReset = new LinkedHashMap<>();
        for (int i = 0; i < e.size(); i++) {
            final String text = procDefContext.params.get(i).getText();
            CatParser.ExpressionContext expressionContext = e.get(i);
            toReset.put(text, paramAssignment.get(text));
            paramAssignment.put(
                    text,
                    expressionContext.accept(this));
        }
        final MCMRelation accept = procDefContext.body.accept(this);
        toReset.forEach((s, mcmRelation) -> {
            if(mcmRelation == null) paramAssignment.remove(s);
            else paramAssignment.put(s, mcmRelation);
        });
        return accept;
    }

    @Override
    public MCMRelation visitExprTryWith(CatParser.ExprTryWithContext ctx) {
        return ctx.e.accept(this);
    }

    @Override
    public MCMRelation visitExprFunctionCall(CatParser.ExprFunctionCallContext ctx) {
        final CatParser.FunctionDefContext functionDefContext = functions.get(ctx.NAME().getText());
        checkNotNull(functionDefContext, "Function with name " + ctx.NAME().getText() + " does not exist.");
        List<CatParser.ExpressionContext> e = ctx.e;
        final Map<String, MCMRelation> toReset = new LinkedHashMap<>();
        for (int i = 0; i < e.size(); i++) {
            final String text = functionDefContext.params.get(i).getText();
            CatParser.ExpressionContext expressionContext = e.get(i);
            toReset.put(text, paramAssignment.get(text));
            paramAssignment.put(
                    text,
                    expressionContext.accept(this));
        }
        final MCMRelation accept = functionDefContext.e.accept(this);
        toReset.forEach((s, mcmRelation) -> {
            if(mcmRelation == null) paramAssignment.remove(s);
            else paramAssignment.put(s, mcmRelation);
        });
        return accept;
    }

    @Override
    public MCMRelation visitLetDefinition(CatParser.LetDefinitionContext ctx) {
        String name = ctx.NAME().getText();
        CatParser.ExpressionContext eCtx = ctx.e;

        relations.remove(name);
        for (CatParser.LetAndDefinitionContext letAndDefinitionContext : ctx.letAndDefinition()) {
            name = letAndDefinitionContext.NAME().getText();
            relations.remove(name);
        }
        MCMRelation rel = eCtx.accept(this);
        final MCMRelation relation = getOrCreateRelation(ctx.NAME().getText(), rel.getArity());
        relation.addRule(new Self(rel));
        for (CatParser.LetAndDefinitionContext letAndDefinitionContext : ctx.letAndDefinition()) {
            name = letAndDefinitionContext.NAME().getText();
            eCtx = letAndDefinitionContext.e;
            rel = eCtx.accept(this);

            final MCMRelation relationAnd = getOrCreateRelation(name, rel.getArity());
            relationAnd.addRule(new Self(rel));
        }

        return relation;
    }

    @Override
    public MCMRelation visitExprCartesian(CatParser.ExprCartesianContext ctx) {
        MCMRelation rel = ctx.e1.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, 2);
        relation.addRule(new CartesianProduct(rel, ctx.e2.accept(this)));
        return relation;
    }

    @Override
    public MCMRelation visitExprBasic(CatParser.ExprBasicContext ctx) {
        return getOrCreateRelation(ctx.NAME().getText(), isUpperCase(ctx.NAME().getText().charAt(0)) ? 1 : 2);
    }

    @Override
    public MCMRelation visitExprMinus(CatParser.ExprMinusContext ctx) {
        MCMRelation rel = ctx.e1.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new Difference(rel, ctx.e2.accept(this)));
        return relation;
    }

    @Override
    public MCMRelation visitExprUnion(CatParser.ExprUnionContext ctx) {
        MCMRelation rel = ctx.e1.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new Union(rel, ctx.e2.accept(this)));
        return relation;
    }

    @Override
    public MCMRelation visitExprComposition(CatParser.ExprCompositionContext ctx) {
        MCMRelation rel = ctx.e1.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new Sequence(rel, ctx.e2.accept(this)));
        return relation;
    }

    @Override
    public MCMRelation visitExprIntersection(CatParser.ExprIntersectionContext ctx) {
        MCMRelation rel = ctx.e1.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new Intersection(rel, ctx.e2.accept(this)));
        return relation;
    }

    @Override
    public MCMRelation visitExprTransitive(CatParser.ExprTransitiveContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new TransitiveClosure(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExprComplement(CatParser.ExprComplementContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new Complement(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExprInverse(CatParser.ExprInverseContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new Inverse(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExprTransRef(CatParser.ExprTransRefContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new ReflexiveTransitiveClosure(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExpr(CatParser.ExprContext ctx) {
        return ctx.e.accept(this);
    }

    @Override
    public MCMRelation visitExprOptional(CatParser.ExprOptionalContext ctx) {
        MCMRelation rel = ctx.e.accept(this);
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, rel.getArity());
        relation.addRule(new IdentityClosure(rel));
        return relation;
    }

    @Override
    public MCMRelation visitExprNull(CatParser.ExprNullContext ctx) {
        final MCMRelation relation = getOrCreateRelation("rule" + ruleNameCnt++, 1);
        relation.addRule(new EmptyRelation());
        return relation;
    }

    @Override
    public String toString() {
        return mcm.toString();
    }
}

File: subprojects/xcfa/cat/src/main/java/hu/bme/mit/theta/cat/dsl/CatVisitor.kt
/*
 *  Copyright 2023 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.cat.dsl

import com.google.common.base.Preconditions
import hu.bme.mit.theta.cat.dsl.gen.CatBaseVisitor
import hu.bme.mit.theta.cat.dsl.gen.CatParser
import hu.bme.mit.theta.cat.dsl.gen.CatParser.*
import hu.bme.mit.theta.graphsolver.patterns.constraints.*
import hu.bme.mit.theta.graphsolver.patterns.patterns.*
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException

class CatVisitor(file: File) : CatBaseVisitor<GraphPattern>() {

    private var mcm: MutableList<GraphConstraint>? = null
    private val ruleNameCnt = 0
    private val file: File
    private val relations: MutableMap<String, GraphPattern>
    private val functions: MutableMap<String, FunctionDefContext>
    private val procedures: MutableMap<String, ProcDefContext>
    private val paramAssignment: MutableMap<String, GraphPattern>

    init {
        relations = LinkedHashMap()
        procedures = LinkedHashMap()
        functions = LinkedHashMap()
        paramAssignment = LinkedHashMap()
        this.file = file
    }

    fun getMcm(): Collection<GraphConstraint>? {
        return mcm
    }

    private fun getOrCreateRelation(name: String, arity: Int): GraphPattern? {
        if (paramAssignment.containsKey(name)) {
            return paramAssignment[name]
        }
        if (relations.containsKey(name)) {
            return relations[name]
        }
        return if (arity == 1) BasicEventSet(name) else BasicRelation(name)
    }

    override fun visitMcm(ctx: McmContext): GraphPattern {
        mcm = ArrayList()
        val file = File(file.parent + File.separator + "stdlib.cat")
        visitIncludedFile(file)
        return super.visitMcm(ctx) ?: EmptyRel
    }

    private fun visitIncludedFile(file: File) {
        if (file.exists() && file.isFile) {
            try {
                val context = CatDslManager.setupCatAntlr(file)
                context.scopeBody().accept(this)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        } else throw RuntimeException(FileNotFoundException(file.absolutePath))
    }

    override fun visitIncludeFile(ctx: IncludeFileContext): GraphPattern {
        val file = File(file.parent + File.separator + ctx.file.text)
        visitIncludedFile(file)
        return EmptyRel
    }

    override fun visitAcyclicDefinition(ctx: AcyclicDefinitionContext): GraphPattern {
        val rule = ctx.e.accept(this) as EdgePattern
        mcm!!.add(if (ctx.negate == null) Acyclic(rule) else Cyclic(rule))
        return EmptyRel
    }

    override fun visitIrreflexiveDefinition(ctx: IrreflexiveDefinitionContext): GraphPattern {
        val rule = ctx.e.accept(this) as EdgePattern
        mcm!!.add(if (ctx.negate == null) Irreflexive(rule) else Reflexive(rule))
        return EmptyRel
    }

    override fun visitEmptyDefinition(ctx: EmptyDefinitionContext): GraphPattern {
        val rule = ctx.e.accept(this)
        mcm!!.add(if (ctx.negate == null) Empty(rule) else Nonempty(rule))
        return EmptyRel
    }

    override fun visitFunctionDef(ctx: FunctionDefContext): GraphPattern {
        functions[ctx.n.text] = ctx
        return EmptyRel
    }

    override fun visitProcDef(ctx: ProcDefContext): GraphPattern {
        procedures[ctx.n.text] = ctx
        return EmptyRel
    }

    override fun visitExprToid(ctx: ExprToidContext): GraphPattern {
        val rel = ctx.e.accept(this) as NodePattern
        return Toid(rel)
    }

    override fun visitExprRange(ctx: ExprRangeContext): GraphPattern {
        val rel = ctx.e.accept(this) as EdgePattern
        return Range(rel)
    }

    override fun visitExprDomain(ctx: ExprDomainContext): GraphPattern {
        val rel = ctx.e.accept(this) as EdgePattern
        return Domain(rel)
    }

    override fun visitProcCall(ctx: ProcCallContext): GraphPattern {
        val procDefContext = procedures[ctx.NAME().text]
        Preconditions.checkNotNull(procDefContext, "Procedure with name " + ctx.NAME().text + " does not exist.")
        val e = ctx.params
        val toReset: MutableMap<String, GraphPattern?> = LinkedHashMap()
        for (i in e.indices) {
            val text = procDefContext!!.params[i].text
            val expressionContext = e[i]
            toReset[text] = paramAssignment[text]
            paramAssignment[text] = expressionContext.accept(this)
        }
        val accept = procDefContext!!.body.accept(this)
        toReset.forEach { (s: String, graphPattern: GraphPattern?) ->
            if (graphPattern == null) paramAssignment.remove(s) else paramAssignment[s] = graphPattern
        }
        return accept
    }

    override fun visitExprTryWith(ctx: ExprTryWithContext): GraphPattern {
        return ctx.e.accept(this)
    }

    override fun visitExprFunctionCall(ctx: ExprFunctionCallContext): GraphPattern {
        val functionDefContext = functions[ctx.NAME().text]
        Preconditions.checkNotNull(functionDefContext, "Function with name " + ctx.NAME().text + " does not exist.")
        val e = ctx.e
        val toReset: MutableMap<String, GraphPattern?> = LinkedHashMap()
        for (i in e.indices) {
            val text = functionDefContext!!.params[i].text
            val expressionContext = e[i]
            toReset[text] = paramAssignment[text]
            paramAssignment[text] = expressionContext.accept(this)
        }
        val accept = functionDefContext!!.e.accept(this)
        toReset.forEach { (s: String, graphPattern: GraphPattern?) ->
            if (graphPattern == null) paramAssignment.remove(s) else paramAssignment[s] = graphPattern
        }
        return accept
    }

    override fun visitLetDefinition(ctx: LetDefinitionContext): GraphPattern {
        var name = ctx.NAME().text
        var eCtx = ctx.e
        relations.remove(name)
        for (letAndDefinitionContext in ctx.letAndDefinition()) {
            name = letAndDefinitionContext.NAME().text
            relations.remove(name)
        }
        var rel = eCtx.accept(this)
        rel.patternName = ctx.NAME().text
        relations[ctx.NAME().text] = rel
        for (letAndDefinitionContext in ctx.letAndDefinition()) {
            name = letAndDefinitionContext.NAME().text
            eCtx = letAndDefinitionContext.e
            rel = eCtx.accept(this)
            relations[name] = rel
        }
        return rel
    }

    override fun visitExprCartesian(ctx: ExprCartesianContext): GraphPattern {
        val rel1 = ctx.e1.accept(this) as NodePattern
        val rel2 = ctx.e2.accept(this) as NodePattern
        return CartesianProduct(rel1, rel2)
    }

    override fun visitExprBasic(ctx: ExprBasicContext): GraphPattern {
        return getOrCreateRelation(ctx.NAME().text, if (Character.isLowerCase(ctx.NAME().text[0])) 2 else 1)!!
    }

    override fun visitExprMinus(ctx: ExprMinusContext): GraphPattern {
        val rel1 = ctx.e1.accept(this)
        val rel2 = ctx.e2.accept(this)
        return if (rel1 is EdgePattern && rel2 is EdgePattern)
            Difference(rel1, rel2)
        else if (rel1 is NodePattern && rel2 is NodePattern)
            DifferenceNode(rel1, rel2)
        else error("Mismatched types")
    }

    override fun visitExprUnion(ctx: ExprUnionContext): GraphPattern {
        val rel1 = ctx.e1.accept(this)
        val rel2 = ctx.e2.accept(this)
        return if (rel1 is EdgePattern && rel2 is EdgePattern)
            Union(rel1, rel2)
        else if (rel1 is NodePattern && rel2 is NodePattern)
            UnionNode(rel1, rel2)
        else error("Mismatched types")
    }

    override fun visitExprComposition(ctx: ExprCompositionContext): GraphPattern {
        val rel1 = ctx.e1.accept(this) as EdgePattern
        val rel2 = ctx.e2.accept(this) as EdgePattern
        return Sequence(rel1, rel2)
    }

    override fun visitExprIntersection(ctx: ExprIntersectionContext): GraphPattern {
        val rel1 = ctx.e1.accept(this)
        val rel2 = ctx.e2.accept(this)
        return if (rel1 is EdgePattern && rel2 is EdgePattern)
            Intersection(rel1, rel2)
        else if (rel1 is NodePattern && rel2 is NodePattern)
            IntersectionNode(rel1, rel2)
        else error("Mismatched types")
    }

    override fun visitExprTransitive(ctx: ExprTransitiveContext): GraphPattern {
        val rel = ctx.e.accept(this) as EdgePattern
        return TransitiveClosure(rel)
    }

    override fun visitExprComplement(ctx: ExprComplementContext): GraphPattern {
        val rel = ctx.e.accept(this)
        return if (rel is EdgePattern) Complement(rel) else if (rel is NodePattern) ComplementNode(rel) else error(
            "Mismatched types")
    }

    override fun visitExprInverse(ctx: ExprInverseContext): GraphPattern {
        val rel = ctx.e.accept(this) as EdgePattern
        return Inverse(rel)
    }

    override fun visitExprTransRef(ctx: ExprTransRefContext): GraphPattern {
        val rel = ctx.e.accept(this) as EdgePattern
        return ReflexiveTransitiveClosure(rel)
    }

    override fun visitExpr(ctx: CatParser.ExprContext): GraphPattern {
        return ctx.e.accept(this)
    }

    override fun visitExprOptional(ctx: ExprOptionalContext): GraphPattern {
        val rel = ctx.e.accept(this) as EdgePattern
        return IdentityClosure(rel)
    }

    override fun visitExprNull(ctx: ExprNullContext): GraphPattern {
        return EmptyRel
    }

    override fun toString(): String {
        return mcm.toString()
    }
}